### PR TITLE
feat(#4): support metrics reporting

### PR DESCRIPTION
**Pull request by h2non**
Still needs some features and better test coverage.

This PR basically supports: #3, #4 and initial implementation of InfluxDB reporter.

### CODE REVIEW

**Comment by h2non** [comment]
TravisCI is failing due `metrics` package is not available in `master`, therefore `go get` cannot fetch it.

File: README.md [hunk 1]
```diff
@@ -50,6 +107,48 @@ func main() {
 }
 ```
 
+## Writting reporters
+
+`metrics` package allows you to write and plug in custom reporters in order to send data to third-party
+metrics and storage providers. 
+
+Reporters most implement the `Reporter` interface, which is a single method:
```
Comment (groyoh) [inline README.md]: `most` -> `must`
`which is` -> `which consists in`

File: README.md [hunk 2]
```diff
@@ -50,6 +107,48 @@ func main() {
...
+func (m *MyCustomReporter) Report(r metrics.Report) {
+  // Print maps
+  fmt.Printf("Counters: %#v \n", r.Counters)
+  fmt.Printf("Gauges: %#v \n", r.Gauges)
+  
+  // Here you should usually map and transform the metrics report
+  // into reporter specific data structures.
+  data := mapReport(r)
+    
+  // Finally, send the metrics, tipically via network to another server
```
Comment (groyoh) [inline README.md]: `// Finally sends the metrics, typically via network to another server`

File: _examples/custom/custom.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	err := vs.Listen()
+	if err != nil {
+		fmt.Errorf("Error: %s\n", err)
+	}
+}
+
+// Simple stub collector
+type collector func(metrics.Reporter)
+
+func (c collector) Collect(r metrics.Reporter) {
```
Comment (groyoh) [inline _examples/custom/custom.go]: Shouldn't this be `Report`?

File: _examples/custom/custom.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+	fmt.Printf("Server listening on port: %d\n", port)
+	err := vs.Listen()
+	if err != nil {
+		fmt.Errorf("Error: %s\n", err)
+	}
+}
+
+// Simple stub collector
+type collector func(metrics.Reporter)
```
Comment (groyoh) [inline _examples/custom/custom.go]: Took me some time to understand how this example works. I think it may be easier to understand if you used a simple struct instead of a function.

File: README.md [hunk 5]
```diff
@@ -50,6 +107,48 @@ func main() {
...
+Reporters must implement the `Reporter` interface, which consists is a single method:
+
+```go
+type Reporter interface {
+  Report(Report)
+}
+```
+
+The metrics publisher will call the `Report` method passing the `Report` struct, which exports 
+two fields containing the counters and gauges as a simple map data type.
```
Comment (groyoh) [inline README.md]: > which exports two fields containing the counters and gauges

could be rephrased to:

> which exports the fields  `Counters` and `Gauges`

File: _examples/custom/custom.go (hunk 3 above)
Comment (h2non) [inline _examples/custom/custom.go]: Must be, actually. Changed.

File: README.md (hunk 5 above)
Comment (h2non) [inline README.md]: LGTM. Changed.

File: meters.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	m.Counter("req.total").Add()
+}
+
+// MeterResponseStatus is used to count the response status code by range (2xx, 4xx, 5xx).
+func MeterResponseStatus(i *Info, m *Metrics) {
+	s := i.Status / 100
+	if s >= 2 && s < 4 {
+		m.Counter("res.status.ok").Add()
+	}
+	if s == 5 {
```
Comment (groyoh) [inline meters.go]: What about using `else if`?

File: meters.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+// MeterRequestOperation is used to count the number of request by HTTP operation.
+// Operation in inferred by HTTP verb:
+//
+// - GET, HEAD = read operation
+// - POST, PUT, PATCH, DELETE = write operation
+func MeterRequestOperation(i *Info, m *Metrics) {
+	if i.Request.Method == "GET" || i.Request.Method == "HEAD" {
+		m.Counter("req.reads").Add()
+	}
```
Comment (groyoh) [inline meters.go:45]: Could also be a `else if` ;)

File: meters.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	if i.Request.Method == "POST" || i.Request.Method == "PUT" ||
+		i.Request.Method == "PATCH" || i.Request.Method == "DELETE" {
+		m.Counter("req.writes").Add()
+	}
+}
+
+// MeterResponseTime is used to measure the HTTP request/response time.
+// Data will be stored in a histogram.
+func MeterResponseTime(i *Info, m *Metrics) {
+	resTime := i.TimeEnd.Sub(i.TimeStart).Nanoseconds() / int64(time.Millisecond)
```
Comment (groyoh) [inline meters.go:56]: What's the resulting type here? float or int?

File: meters.go (hunk 7 above)
Comment (h2non) [inline meters.go:45]: Syntactically speaking is cleaner for me in this way. The cost of the operation in terms of performance is irrelevant in this case, but I can change it, don't mind too much actually.

File: meters.go (hunk 8 above)
Comment (h2non) [inline meters.go:56]: Go type system will implicitly cast it as `int64`, I guess.

File: writer.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+// Header implements http.ResponseWriter Header method.
+func (l *metricWriter) Header() http.Header {
+	return l.w.Header()
+}
+
+// WriteHeader implements http.ResponseWriter WriteHeader method.
+func (l *metricWriter) WriteHeader(code int) {
+	if l.info.Status != 0 {
```
Comment (groyoh) [inline writer.go:33]: Can this happen?

File: runtime.go [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+	// statistics and the Start function should return. If Done is set, upon shutdown
+	// all gauges will be sent a final zero value to reset their values to 0.
+	Done <-chan struct{}
+
+	// gaugeFunc stores the function used to report stats.
+	gaugeFunc GaugeFunc
+}
+
+// NewRuntimeCollector creates a new RuntimeCollector that will periodically output statistics to gaugeFunc. It
+// will aslo set the values of the exported fields to the described defaults. The values
```
Comment (groyoh) [inline runtime.go]: typo at `aslo`

File: runtime_test.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+		c.Start()
+	}()
+	time.Sleep(1500 * time.Millisecond)
+	close(done)
+	<-collectorShutdown
+
+	// we're going to check a few keys to make sure that the stats are being output
+	// correctly. We expect there to be three of each key because there is one
+	// from the initial stats, plus another one after the 1 second pause time, plus
+	// the final zeroing out on shutdown
```
Comment (groyoh) [inline runtime_test.go:37]: Maybe this test could also test that the latest one contains zero values.

File: writer.go (hunk 9 above)
Comment (h2non) [inline writer.go:33]: Unfortunately yes. Go `net/http` prints a warning in `stdout`, and I've seen it multiple times.

File: meters.go (hunk 7 above)
Comment (groyoh) [inline meters.go:45]: It's not about about performance but more about the fact that when I read `else if` I immediately understand that the two case are mutually exclusive.

File: meters.go (hunk 8 above)
Comment (groyoh) [inline meters.go:56]: So we will never have sub-millisecond metrics, right?

File: reporters/influx/influx.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+// New creates a new InfluxDB reporter which will post the metrics to the specified server.
+func New(c Config) *Reporter {
+	re := &reporter{config: c}
+	if err := re.makeClient(); err != nil {
+		log.Printf("unable to make InfluxDB client. err=%v", err)
+	}
+	return re
+}
+
+// Report implements the metrisc.Reporter interface.
```
Comment (groyoh) [inline reporters/influx/influx.go]: typo at `metrisc.Reporter`

File: reporters/influx/influx.go [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+func (r *Reporter) Report(re metrics.Report) {
+	if err := r.send(re); err != nil {
+		log.Printf("infludb: error sending metrics err=%v", err)
+	}
+}
+
+func (r *Reporter) makeClient() (err error) {
+	r.client, err = client.NewHTTPClient(client.HTTPConfig{
```
Comment (groyoh) [inline reporters/influx/influx.go:51]: What's the rule of thumb for having a named return value?

File: meters.go (hunk 8 above)
Comment (h2non) [inline meters.go:56]: Correct.

File: reporters/influx/influx.go [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+type Reporter struct {
+	config Config
+	client client.Client
+}
+
+// New creates a new InfluxDB reporter which will post the metrics to the specified server.
+func New(c Config) *Reporter {
+	re := &reporter{config: c}
+	if err := re.makeClient(); err != nil {
+		log.Printf("unable to make InfluxDB client. err=%v", err)
```
Comment (groyoh) [inline reporters/influx/influx.go:36]: Shouldn't this return an error instead? When it cannot make a client, subsequent call to `Report` will break, right?

File: metrics.go [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+	if !ok {
+		counter = metrics.Counter(key)
+		m.counters[key] = counter
+	}
+	return counter
+}
+
+// Guage returns a guide metric by key.
+// If the guige doesn't exists, it will be transparently created.
+func (m *Metrics) Guage(key string) metrics.Gauge {
```
Comment (groyoh) [inline metrics.go:72]: Any reason why it is called `Guage`? Name conflicts?

File: metrics.go [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+	counter, ok := m.counters[key]
+	if !ok {
+		counter = metrics.Counter(key)
+		m.counters[key] = counter
+	}
+	return counter
+}
+
+// Guage returns a guide metric by key.
+// If the guige doesn't exists, it will be transparently created.
```
Comment (groyoh) [inline metrics.go]: typo `guige`

File: metrics.go [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+	if !ok {
+		hist = metrics.NewHistogram(key, 0, 1e8, 5)
+		m.histograms[key] = hist
+	}
+	return hist
+}
+
+// Snapshot collects and returns a report of the existent counters and gauges metrics
+// to be consumed by metrics publishers and listeners.
+func (m *Metrics) Snapshot() Report {
```
Comment (groyoh) [inline metrics.go:98]: I don't get this method. Shouldn't it be:

``` go
func (m *Metrics) Snapshot() Report {
  m.Lock()
  defer m.Unlock() 
  return Report{Gauges: m.gauges, Counters: m.counters}
}
```

Actually I'm not sure if that would actually be a snapshot. Are `map` referenced by pointers or value?

File: meter.go [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+	"net/http"
+	"sync"
+	"time"
+
+	"gopkg.in/vinxi/layer.v0"
+)
+
+// PublishInterval defines the amount of time to wait between metrics publish cycles.
+// Defaults to 15 seconds.
+var PublishInterval = 15 * time.Second
```
Comment (groyoh) [inline meter.go:13]: Shouldn't this be customizable?

File: metrics.go [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+	c, g := metrics.Snapshot()
+	return Report{Gauges: g, Counters: c}
+}
+
+// Reset resets all the metrics (counters, gauges & histograms) to zero.
+// You should collect them first with Snapshop(), otherview the collected data will be lost.
+func (m *Metrics) Reset() {
+	metrics.Reset()
```
Comment (groyoh) [inline metrics.go:106]: Could you explain this?

File: metrics.go [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+
+// Snapshot collects and returns a report of the existent counters and gauges metrics
+// to be consumed by metrics publishers and listeners.
+func (m *Metrics) Snapshot() Report {
+	c, g := metrics.Snapshot()
+	return Report{Gauges: g, Counters: c}
+}
+
+// Reset resets all the metrics (counters, gauges & histograms) to zero.
+// You should collect them first with Snapshop(), otherview the collected data will be lost.
```
Comment (groyoh) [inline metrics.go]: Typos here `Snapshop()` :smile:  and `otherview`.

File: meter.go [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+// Stop stops the publish goroutine.
+func (m *Meter) Stop() {
+	close(m.quit)
+	// close(m.runtime.Done) // TODO
+}
+
+// measureHTTP instruments and logs an incoming HTTP request and response.
+func (m *Meter) measureHTTP(h http.Handler) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		m := newMetricsWriter(w, r, m.gauge)
```
Comment (groyoh) [inline meter.go]: `m` is kind of weird as variable name for a `ReponseWriter`

File: meter.go [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+
+// Register registers the metrics middleware function.
+func (m *Meter) Register(mw layer.Middleware) {
+	mw.UsePriority("request", layer.TopHead, m.measureHTTP)
+}
+
+// Publish release store.
+func (m *Meter) Publish() {
+	for _, reporter := range m.reporters {
+		reporter.Report(m.metrics.Snapshot())
```
Comment (groyoh) [inline meter.go]: Why not snapshot once before the loop?

File: meter.go [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+	}
+	m.metrics.Reset()
```
Comment (groyoh) [inline meter.go]: Between the snapshot and the time it take the reporter to report the metrics, we may track some values that will be discarded, right? Shouldn't it be somehow `SnapshotAndReset`?

**Comment by groyoh** [comment]
I finished my first round of review. It was quite hard to grasp everything so I'd like to review it again tomorrow.  @h2non would that be fine with you?

File: reporters/influx/influx.go (hunk 13 above)
Comment (h2non) [inline reporters/influx/influx.go:51]: Basically, convenience and to avoid explicit local variable declarations and boilerplate, especially in multiple assignment. E.g:

``` go
func (r *Reporter) makeClient() (err error) {
    r.client, err = client.NewHTTPClient(client.HTTPConfig{
        Addr:     r.config.URL,
        Username: r.config.Username,
        Password: r.config.Password,
    })
    return
}
```

Versus:

``` go
func (r *Reporter) makeClient() error {
     var err error
     r.client, err = client.NewHTTPClient(client.HTTPConfig{
        Addr:     r.config.URL,
        Username: r.config.Username,
        Password: r.config.Password,
    })
    return err
}
```

Versus:

``` go
func (r *Reporter) makeClient() error {
     client, err := client.NewHTTPClient(client.HTTPConfig{
        Addr:     r.config.URL,
        Username: r.config.Username,
        Password: r.config.Password,
    })
    r.client = client
    return err
}
```

File: metrics.go (hunk 15 above)
Comment (h2non) [inline metrics.go:72]: I think it's the correct idiom.

File: reporters/influx/influx.go (hunk 14 above)
Comment (h2non) [inline reporters/influx/influx.go:36]: `New()` is designed to be consumed directly by `metrics.New(...Reporter)`. 
Having multiple return values makes that impossible, but handling errors is important.

> When it cannot make a client, subsequent call to Report will break, right?

Should fail with a `nil` reference error.

> How to improve this?

Error handling should be a priority, and if something fails, it should make noise. 

I just modified the `Reporter` interface to return an error, making possible to deal with that programmaticaly in the future from the `metrics` package (perhaps binding some error-specific handler):

``` go
type Reporter interface {
    Report(Report) error
}
```

What do you think? Other ideas?

File: metrics.go (hunk 16 above)
Comment (h2non) [inline metrics.go]: Fixed.

File: metrics.go (hunk 17 above)
Comment (h2non) [inline metrics.go:98]: Metrics fields are not only stored in this package. They're stored in `codahale/metrics`, more specifically [here](https://github.com/codahale/metrics/blob/master/metrics.go#L302-L306). 

`map` and `slices` technically are first-class values in Go. `map` is actually mutable, so behaves like a pointer. `slices` aren't.

On every `Snapshot` calls, a new clone of the original map is created, providing a side-effects free map structure. See code [here](https://github.com/codahale/metrics/blob/master/metrics.go#L173-L185).

We could stop relying in `codahale/metrics` to retrieve metrics in the future.

File: meter.go (hunk 18 above)
Comment (h2non) [inline meter.go:13]: It is, actually. You can do: `metrics.PublishInterval = 5 * time.Second`.

I would like to support a `Config` struct to pass instance specific configuration, maybe in the next bunch of improvements.

File: metrics.go (hunk 19 above)
Comment (h2non) [inline metrics.go:106]: See: https://github.com/codahale/metrics/blob/master/metrics.go#L140-L156

File: meter.go (hunk 18 above)
Comment (groyoh) [inline meter.go:13]: Ah yeah, true!

File: meter.go (hunk 21 above)
Comment (h2non) [inline meter.go]: Agree.

File: meter.go (hunk 23 above)
Comment (h2non) [inline meter.go]: `Snapshot()` is not synchronized via mutex, so technically snapshots could not be consistent across concurrent calls, but this is mostly a theoretical scenario because this shouldn't happen since calls to `Publish()` are done by one single goroutine that as a time window interval, making it sequential.

Alternatively, each `Report()` call could be called in its own goroutine. I think I'm going to do some changes regarding to this.

File: meter.go (hunk 22 above)
Comment (h2non) [inline meter.go]: The same I described before. Initially I thought would be better to provide a side effects free `map` to the reporters, but this could be reverted. I'm not sure how to design to avoid nested issues in third-party developers.

File: metrics.go (hunk 20 above)
Comment (h2non) [inline metrics.go]: Fixed.

**Comment by h2non** [comment]
@groyoh I would like to merge this now just to have something usable in `master`, but I'll keep the branch so tomorrow we can continue with the code review with new improvements.

In any case, I did all the fixes and supported most of your proposals, also with further improvements.
